In [2]:
import pandas as pd 
from scipy.stats import norm
import numpy as np

In [3]:
pip install xlrd

Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\Phillip\AppData\Local\Programs\Python\Python39\python.exe -m pip install --upgrade pip' command.


In [24]:
returns_df = pd.read_excel('Problem Set 1 Data.xls', index_col = 0) + 1 


In [27]:
def portfolio_replication(
        returns,
        S_0 = 100,
        K = 90,
        r = 0.0,
        vol= 0.25,
        Total_Weeks = 20,
        asset = 'Return series 1'
):
    if isinstance(returns, pd.DataFrame):
        weekly_returns = returns[asset].copy()
    else:
        weekly_returns = returns.copy()
    
    #Create initial conditions for price process
    initial = pd.Series([1.0], index = [-1]) 
    #Use ignore_index to get around indexing issues. 
    all_returns = pd.concat([initial, weekly_returns], ignore_index = True)
    
    #Convert returns to stock price - starting from Week 0 - use .iloc to get rid of index = -1 item
    prices = (all_returns.cumprod() * S_0).iloc[1:]

    def BSM(S, tau, K = K, vol = vol, r = r):

        #Put value is just payoff at maturity
        if tau <= 0:
            put_price = np.max(K - S, 0)
            return put_price, 0 
        
        d1 = (np.log(S / K) + (r + 0.5 * vol**2 )*(tau)) / (vol * np.sqrt(tau))
        d2 = d1 - vol * np.sqrt(tau)
        put_price = K * np.exp( r * tau) * norm.cdf(-d2) - S * norm.cdf(-d1)
        delta = norm.cdf(d1) - 1 

        return put_price, delta
        
    
    ''' Creating Dataframe to store replicating portfolio values'''

    columns = ['Week', 'S_t', 'Time-2-Mat', 'BSM Price', 'Delta'
               , 'dDelta', 'Hedge', 'Cash Bal', 'Port Val']
    
    df = pd.DataFrame(index = np.arange(0, Total_Weeks + 1), columns = columns)

    ''' Creating the initial portfolio 
        - Cash position = Put Option - Delta * S 
        - Cash position is from shorting - Delta number of Shares 

    '''
    tau0 = Total_Weeks/52 
    P0, delta0 = BSM(S_0, tau = tau0)
    #Cash Position at t = 0
    cash0 = P0 - delta0 * S_0 

    df.loc[0] = [
        0, #Week
        S_0, #S_t
        tau0, #Time2Mat
        P0, #BSM Price
        delta0, #Delta
        np.nan, #dDelta 
        np.nan, #rebalancing trade 
        cash0, #cash position
        cash0 + delta0*S_0, #port val 
                ]
    
    ''' Step Forward : For Loop Replication '''
    Delta_prior_i = delta0
    cash_prior = cash0
    for period in range(1, Total_Weeks + 1):
        #Share Price in Period i 
        S_i = prices.loc[period]
        #New time to maturity
        tau_i = (Total_Weeks - period)/ 52
        #New Put Price
        Put_i, Delta_i = BSM(S_i, tau_i)
        #New Delta
        dDelta = Delta_i - Delta_prior_i
        #Hedge Trade
        hedge = -dDelta * S_i 
        #New Cash Balance
        cash_i = cash_prior + hedge 
        #Port Val
        port_val_i = cash_i + Delta_i * S_i 

        #Save to dataframe
        df.loc[period] = [period, S_i, tau_i, Put_i, Delta_i, dDelta, hedge, cash_i, port_val_i]

        #Reset Values
        Delta_prior_i = Delta_i
        cash_prior = cash_i 
    

    return df


df = portfolio_replication(returns = returns_df)
df2 = portfolio_replication(returns = returns_df, asset = 'Return series 2')

In [29]:
df

Week        S_t Time-2-Mat  BSM Price     Delta    dDelta      Hedge  \
0     0        100   0.384615   2.171644 -0.224502       NaN        NaN   
1     1  91.499655   0.365385   4.748612 -0.426648 -0.202146  18.496305   
2     2  88.283613   0.346154   6.128892 -0.522873 -0.096225   8.495105   
3     3  85.742017   0.326923   7.420173 -0.605493  -0.08262   7.083991   
4     4  88.233645   0.307692   5.859624 -0.529334  0.076159  -6.719793   
5     5  88.209642   0.288462   5.717861  -0.53288 -0.003546   0.312793   
6     6  89.557747   0.269231   4.867292 -0.489276  0.043604  -3.905093   
7     7  87.925556       0.25   5.547532 -0.549363 -0.060088   5.283233   
8     8  86.937508   0.230769   5.941933 -0.590263   -0.0409   3.555717   
9     9  86.822846   0.211538   5.838087 -0.600667 -0.010404    0.90331   
10   10  89.308684   0.192308   4.274582 -0.506191  0.094476  -8.437569   
11   11  91.468279   0.173077   3.074367 -0.417772  0.088418  -8.087468   
12   12  92.872924   0.153846   2.320914 -0.355886  0.061886  -5.747559   
13   13  91.991882   0.134615   2.426978 -0.388007 -0.032121   2.954826   
14   14  91.604254   0.115385   2.339466 -0.401095 -0.013088    1.19895   
15   15  85.446302   0.096154   5.575293 -0.735982 -0.334887  28.614846   
16   16  82.367621   0.076923   7.916989 -0.893137 -0.157155  12.944523   
17   17  75.091813   0.057692  14.909969 -0.998586 -0.105449    7.91833   
18   18  74.159906   0.038462   15.84013 -0.999956  -0.00137   0.101624   
19   19  71.137364   0.019231  18.862636      -1.0 -0.000044   0.003099   
20   20  68.427302        0.0  21.572698         0       1.0 -68.427302   

     Cash Bal   Port Val  
0   24.621842   2.171644  
1   43.118147   4.079988  
2   51.613252   5.452106  
3   58.697244   6.781039  
4    51.97745   5.272375  
5   52.290243   5.285081  
6    48.38515   4.566703  
7   53.668383   5.365294  
8     57.2241   5.908092  
9    58.12741   5.975772  
10  49.689841   4.482612  
11  41.602373   3.389444  
12  35.854815   2.802622  
13  38.809641   3.116173  
14  40.008591   3.266575  
15  68.623438   5.736499  
16  81.567961   8.002353  
17  89.486291   14.50065  
18  89.587914  15.431239  
19  89.591014   18.45365  
20  21.163712  21.163712

In [30]:
df2

Week        S_t Time-2-Mat  BSM Price     Delta    dDelta      Hedge  \
0     0        100   0.384615   2.171644 -0.224502       NaN        NaN   
1     1    94.3457   0.365385   3.646257 -0.349153 -0.124651   11.76032   
2     2  92.139898   0.346154   4.337019 -0.407764  -0.05861   5.400337   
3     3  90.375418   0.326923   4.953134 -0.459937 -0.052173   4.715198   
4     4  92.122375   0.307692   4.043405 -0.406167   0.05377  -4.953389   
5     5  92.105705   0.288462   3.892882 -0.405406  0.000761  -0.070095   
6     6  93.042051   0.269231   3.366217 -0.374059  0.031347  -2.916576   
7     7  91.914103       0.25   3.639828 -0.408712 -0.034653   3.185096   
8     8  91.227054   0.230769   3.752808 -0.431402 -0.022689   2.069892   
9     9   91.14702   0.211538   3.604095 -0.433437 -0.002035   0.185496   
10   10  92.882915   0.192308   2.719916  -0.36602  0.067417  -6.261929   
11   11  94.376936   0.173077   2.026234 -0.305523  0.060497  -5.709507   
12   12  95.340997   0.153846   1.561419 -0.262079  0.043443  -4.141941   
13   13  94.739366   0.134615   1.524195  -0.27247  -0.01039   0.984382   
14   14  94.473821   0.115385   1.382906 -0.269695  0.002775  -0.262118   
15   15   90.24933   0.096154   2.663668 -0.470327 -0.200632  18.106898   
16   16  88.086324   0.076923   3.536739 -0.608457  -0.13813  12.167324   
17   17  82.910548   0.057692   7.294126 -0.909281 -0.300824  24.941477   
18   18  82.226113   0.038462   7.827935 -0.965468 -0.056187   4.620073   
19   19  79.996881   0.019231  10.003375 -0.999639 -0.034171   2.733584   
20   20   77.96968        0.0   12.03032         0  0.999639 -77.941542   

     Cash Bal  Port Val  
0   24.621842  2.171644  
1   36.382161  3.441045  
2   41.782499  4.211208  
3   46.497696  4.930699  
4   41.544307  4.127209  
5   41.474212   4.13398  
6   38.557636  3.754379  
7   41.742732  4.176299  
8   43.812623  4.457104  
9   43.998119  4.491631  
10  37.736191   3.73923  
11  32.026684  3.192389  
12  27.884743  2.897847  
13  28.869124  3.055521  
14  28.607006  3.127874  
15  46.713904  4.267199  
16  58.881228  5.284519  
17  83.822705  8.433755  
18  88.442778  9.056098  
19  91.176362  11.20835  
20   13.23482  13.23482

# Question A 

BSM Price of Put is $2.171643

In [ ]:
df['BSM Price'][0]

np.float64(2.1716436485757598)

# Question B & C 

Trader's final cash balance is $21.16 dollars which is close to the Black-Sholes price of the put option at $21.57. This shows for this given returns series, the portfolio insurance strategy approximates a put option well. Additionally, the investor paid less to synethetically replicate a put option than outright buying the put. 

Smaller market movements (shown with a lower realised volatility) is beneficial for the vega-exposure that this strategy holds. In this return series, the synthethic put was replicated to be cheaper than the put premium

In [21]:
df.loc[20]

Week                 20
S_t           68.427302
Time-2-Mat          0.0
BSM Price     21.572698
Delta                 0
dDelta              1.0
Hedge        -68.427302
Cash Bal      21.163712
Port Val      21.163712
Name: 20, dtype: object

In [ ]:
(df.loc[20, 'Cash Bal'] - df.loc[20 , 'BSM Price']).round(2)

np.float64(-0.41)

# Question D

Final cash balance is $13.23 dollars. Compared to payoff of only holding the put would yield $12.03 dollars which is $1.20 less.

Realised volatility is 39.5% whereas implied volatility given by the question is 25% (the vol value used to price BSM-model puts). This volatility premium means the put option is 'expensive' as the holder of the protfolio could have. This intuitively makes sense, given implied volatility is essientally how traders see the price as options. Therefore, if implied volatility is higher than realised volatility, then the prices quoted by market makers is overvalued


In [32]:
df2.loc[20]

Week                 20
S_t            77.96968
Time-2-Mat          0.0
BSM Price      12.03032
Delta                 0
dDelta         0.999639
Hedge        -77.941542
Cash Bal       13.23482
Port Val       13.23482
Name: 20, dtype: object

In [ ]:
df2['S_t'].std() * np.sqrt(52)

np.float64(39.47957887722939)